flink-kubernetes/src/main/java/org/apache/flink/kubernetes/kubeclient/resources/KubernetesSharedInformer.java

```diff
@@ -265,14 +265,26 @@ private WatchCallback(
         private void run(Consumer<WatchCallbackHandler<T>> handlerConsumer) {
             if (this.executorService == null) {
                 handlerConsumer.accept(handler);
-            } else {
-                this.executorService.submit(
+                return;
+            }
+            final int present = handling.getAndIncrement();
+            try {
+                queue.put(handlerConsumer);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+                return;
+            }
+            if (present == 0) {
+                executorService.submit(
                         () -> {
-                            callbackLock.lock();
-                            try {
-                                handlerConsumer.accept(handler);
-                            } finally {
-                                callbackLock.unlock();
+                            while (true) {
+                                final Consumer<WatchCallbackHandler<T>> h = queue.poll();
```

Comment:
This makes sense 👍 Have you been able to verify locally that this fixes the issue?

I'm just thinking whether executing callbacks in a loop is ok (does it have potential to block the thread for a long time?), or we should execute callbacks one by one (each callback gets a new Runnable). IMO it's ok as is, just a first thought I had when I was looking at the code.

Reply:
nit: We may want to use a blocking poll here.

Reply:
no need blocking poll, the poll can always get a element.

Reply:
I've verify locally for many rounds, no exception ever found so far.

The execution thread will only survive for a short period only if new event comes before the last event complete.

Reply:
i mean the loop by "The execution thread will only survive"

Reply:
I think there is a rare case where it won't, basically if you poll between `handling.getAndIncrement() > 0` and 
`queue.put(handlerConsumer);`. Also if `queue.put(handlerConsumer)` would get interrupted in case `present` is larger than `0` it may result in infinite busy waiting loop.

Reply:
i think execute callbacks one by one and a new Runnable for each one needs another thread to maintain the order also.

Reply:
> i think execute callbacks one by one and a new Runnable for each one needs another thread to maintain the order also.

BlockingQueue where you store incoming callbacks from a single thread (this maintains order) + lock (doesn't need to be fair, ordering is maintained by the queue, we just need an exclusive access), inside which you poll callback from the queue and execute it, should be enough.

---
Problem with the previous implementation was, that poll from the executor service queue, was not guarded by the lock.

Reply:
Yeah, it's also ok. i think it still some guard to make the single thread enter and exit.

Reply:
So, @dmvk would you mind i keep the current implementation?

Reply:
Personally i find it harder to think and reason about as incrementing number of elements in the queue and actually putting and element into the queue is not atomic.

I'd personally prefer something like:

```
Preconditions.checkState(
      callbackQueue.add(handlerConsumer), "Unable to put callback into a queue.");
executorService.submit(
      () -> {
          callbackLock.lock();
          try {
              Objects.requireNonNull(
                              callbackQueue.poll(), "Callback queue is empty.")
                      .accept(handler);
          } finally {
              callbackLock.unlock();
          }
      });
```

But that being said, I don't think I have a strong objection against your solution. WDYT @wangyang0918 ?

Reply:
@dmvk After more thinking, I now prefer your solution. If the caller uses a shared single-threaded executor to call watch, the current implementation may cause the thread to always be occupied by a single watcher. Your solution can avoid this scenario.